### PR TITLE
Fix LumiCache in AlcaBeamMonitor and OnlineBeamMonitor

### PR DESCRIPTION
**Pull request by francescobrivio**
#### PR description:
Triggered by [this CMSTalk post](https://cms-talk.web.cern.ch/t/issues-with-alcabeammonitor-in-offline-dqm/23945) and as a follow-up of https://github.com/cms-sw/cmssw/issues/37163 and https://github.com/cms-sw/cmssw/issues/39973, @gennai and I implemented the changes suggested by @makortel in https://github.com/cms-sw/cmssw/issues/37163#issuecomment-1061301312.
The changes include:
 - In AlcaBeamMonitor:
    - redefine (and use) the `LuminosityBlockCache`
    - rename the histogram titles from "Scalers" to "Online"
    - Fixed the `Event::getByToken()` try/catch block as suggested by Matti
    - remove the unused `numberOfProcessedLumis_` variable
    - moved the filling of `processedLumis_` to `globalEndLuminosityBlock()` and removed the obsolete `mutable` declaration
- In OnlineBeamMonitor:
   - redefine (and use) the `LuminosityBlockCache`
    - remove the unused `numberOfProcessedLumis_` variable
    - moved the filling of `processedLumis_` to `globalEndLuminosityBlock()` and removed the obsolete `mutable` declaration

@makortel @mmusich please let us know what you think of these changes.

#### PR validation:
~We tested this with wf `138.5` which runs the full `ALCA` reco sequence in step3, using events from multiple lumisections.
I'm not sure if this is the correct way to do it or if some other trick is needed to properly test this, suggestions are welcome.~
See below the discussion about testing this PR properly

#### Backport:
Not a backport. Backports available at:
 - 13_1_X: #41722
 - 13_0_X: #41723

### CODE REVIEW

**Review by makortel — COMMENTED** [review]
Thanks @gennai and @francescobrivio! This is the proper way to use `LuminosityBlockCache` in `edm::one` modules.

File: DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc [hunk 1]
```diff
@@ -224,12 +224,13 @@ void AlcaBeamMonitor::bookHistograms(DQMStore::IBooker& ibooker, edm::Run const&
...
-                                                                                      const EventSetup& iSetup) const {
+std::shared_ptr<alcabeammonitor::BeamSpotInfo> AlcaBeamMonitor::globalBeginLuminosityBlock(
+    const LuminosityBlock& iLumi, const EventSetup& iSetup) const {
   // Always create a beamspot group for each lumi weather we have results or not! Each Beamspot will be of unknown type!
 
-  vertices_.clear();
-  beamSpotsMap_.clear();
+  auto beamSpotInfo = std::make_shared<alcabeammonitor::BeamSpotInfo>();
+  beamSpotInfo->vertices_.clear();
+  beamSpotInfo->beamSpotMap_.clear();
```
[the remark this answers is not in the document]
Comment (francescobrivio) [reply DQM/BeamMonitor/plugins/AlcaBeamMonitor.cc]: done in [f293642](https://github.com/cms-sw/cmssw/pull/41700/commits/f29364240430bfdfc8d1055be62012844e6d7b8d)

File: DQM/BeamMonitor/plugins/AlcaBeamMonitor.h [hunk 2]
```diff
@@ -28,30 +28,34 @@ class BeamFitter;
 class PVFitter;
 
 namespace alcabeammonitor {
-  struct NoCache {};
+  struct BeamSpotInfo {
+    mutable std::vector<reco::VertexCollection> vertices_;
+    typedef std::map<std::string, reco::BeamSpot> BeamSpotContainer;
+    mutable BeamSpotContainer beamSpotMap_;
```
[the remark this answers is not in the document]
Comment (francescobrivio) [reply DQM/BeamMonitor/plugins/AlcaBeamMonitor.h]: done in [f293642](https://github.com/cms-sw/cmssw/pull/41700/commits/f29364240430bfdfc8d1055be62012844e6d7b8d)

**Comment by francescobrivio** [comment]
In the last push I:
 - applied the suggested comments
 - removed the unused `numberOfProcessedLumis_` from both `AlcaBeamMonitor` and `OnlineBeamMonitor`
 - removed the `mutable` from `processedLumis_` and moved its filling to `globalEndLuminosityBlock()` from both `AlcaBeamMonitor` and `OnlineBeamMonitor`, as suggested in https://github.com/cms-sw/cmssw/issues/37163#issuecomment-1061305897
 - applied the same LumiCache fixes to `OnlineBeamMonitor`

**Comment by francescobrivio** [comment]
I also tested this with:
```
runTheMatrix.py -l 138.5 -j 8 --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200
```
which gives, e.g.:
Screenshot 2023-05-16 alle 23 33 25

To be honest, given the tests parameters I was expecting to see 5 lumisections (1000 evts / 200 evts per LS), but well...better than nothing!
I'm not sure how to instruct the bot to use the same options for testing...any suggestion?

**Comment by francescobrivio** [comment]
test parameters:
 - enable = threading

**Comment by francescobrivio** [comment]
@cmsbuild please test

 - while waiting for further feedback on the commits and on the correct way to test them, we can still see if the standard tests work in multithreading

**Comment by mmusich** [comment]
>  also tested this with:
> runTheMatrix.py -l 138.5 -j 8 --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200

as far I can tell, when you run in multi-threaded mode (trading `-nStreams 2` with `-t 4`) the plot you show here is changed with respect to the one obtained in a vanilla `CMSSW_13_2_X_2023-05-16-2300`.
Screenshot from 2023-05-17 12-51-31




On the other hand none of the two is consistent with the observations in the production enviroment (see [CMSTalk post](https://cms-talk.web.cern.ch/t/issues-with-alcabeammonitor-in-offline-dqm/23945), in which apparently only one LS is plotted, https://tinyurl.com/2gtx6s9q). I gather we're not capturing with this command some feature that is running in the Tier-0 jobs.

**Comment by tvami** [comment]
test parameters:
  - workflow = 138.5
  - workflow_options =  --nStreams 2 --nEvents 1000 --numberEventsInLuminosityBlock 200

**Comment by tvami** [comment]
Screen Shot 2023-05-17 at 07 51 49
the bot liked it, so I think this is it

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by francescobrivio** [comment]
actually I believe this specific wf is using some of the modules defined here:

https://github.com/cms-sw/cmssw/blob/96423893105e2482c0882f8f0e5fbc61d745605e/Configuration/AlCa/python/autoAlca.py#L60-L66

for which the concurrent lumis are blocked/forbidden. @makortel can you confirm?

**Comment by mmusich** [comment]
now running
```
 runTheMatrix.py -l 138.5 -j 8 -t 4 --nEvents 1000 --numberEventsInLuminosityBlock 200 --command='--nConcurrentLumis 2'
```

with this branch I get this


Screenshot from 2023-05-17 14-52-29

and things get more consistent (vanilla has only one LS shown), and go in the right direction!

**Comment by makortel** [comment]
The `--numberEventsInLuminosityBlock 200` parameter does have any impact here, because the workflow reads existing files and we don't allow reshuffling events into lumis after the fact.

> actually I believe this specific wf is using some of the modules defined here:
> for which the concurrent lumis are blocked/forbidden. @makortel can you confirm?

Indeed the step3 seems to have `PromptCalibProd` and `PromptCalibProdSiPixelAli`, which cause the step3 being run with 1 concurrent lumi regardless of `--command='--nConcurrentLumis 2'`. The override needs to be done in a way that `ConfigBuilder` does not "see" it, which is basically a customise command (`--command='--customise_commands "process.options.numberOfConcurrentLuminosityBlocks=2"'`)  or function, or editing the generated configuration file by hand. But note that the modules currently explicitly not supporting concurrent lumis will misbehave.

In general (two) concurrent lumis are enabled by default for all jobs that use at least 2 streams (with `ConfigBuilder` setting it back to one if some specific generator or AlCa modules are present).

**Comment by mmusich** [comment]
> The override needs to be done in a way that ConfigBuilder does not "see" it, which is basically a customise command (--command='--customise_commands "process.options.numberOfConcurrentLuminosityBlocks=2"')

I don't see that in my test.
When I run `runTheMatrix.py -l 138.5 -j 8 -t 4 --nEvents 1000 --numberEventsInLuminosityBlock 200 --command='--nConcurrentLumis 2'` and then dump the step 2 configuration:

```
edmConfigDump step2_RAW2DIGI_L1Reco_RECO_ALCAPRODUCER_DQM_ENDJOB.py > dump.py
```

I see:
```
$ more dump.py | grep currentLumi
    numberOfConcurrentLuminosityBlocks = cms.untracked.uint32(2),
```

**Comment by makortel** [comment]
> ump the step 2 configuration

Can you check the step3? (then of course if these modules are used only in the step2, behavior of step3 doesn't matter, but also then the step2 should have concurrent lumis enabled automatically in conjunction of multiple streams)

**Comment by mmusich** [comment]
> Can you check the step3? (then of course if these modules are used only in the step2, behavior of step3 doesn't matter, but also then the step2 should have concurrent lumis enabled automatically in conjunction of multiple streams)

indeed

```
$ edmConfigDump step3_ALCAOUTPUT_ALCA.py  > dump2.py
$ more dump2.py | grep currentLumi
    numberOfConcurrentLuminosityBlocks = cms.untracked.uint32(1),
```

but that has no effect on where the DQM module is run.

**Comment by francescobrivio** [comment]
> is this ready to be tested?

Might benefit of some commit-squashing, but apart from that i don't have any other updates.

**Comment by mmusich** [comment]
> Might benefit of some commit-squashing, but apart from that i don't have any other updates.

standing by for a squash .

**Comment by francescobrivio** [comment]
> > Might benefit of some commit-squashing, but apart from that i don't have any other updates.
> 
> standing by for a squash .

Thanks Marco! Done in the last push:
 - one commit for all the changes in `AlcaBeamMonitor`
 - one commit for all the changes in `OnlineBeamMonitor`

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by tvami** [comment]
> You potentially added 1829 lines to the logs

was this expected?

**Comment by mmusich** [comment]
> was this expected?

Given that in the backport tests there is no such increase (
https://github.com/cms-sw/cmssw/pull/41722#issuecomment-1553130713) I tend to conclude it's coming from the non standard test parameters used here.

**Comment by perrotta** [comment]
The extra lines in the log are due to the fact the wf 138.5_ExpressCollisions2021 runs with 100 events in the baseline and 1000 events in these tests: is it because of the multithreading enabled?
(Not an issue for the content of this PR, in any case)

**Comment by francescobrivio** [comment]
> is it because of the multithreading enabled?

The multithreading and 100 events were part of the non-standard parameters set for the tests (my fault).

@tvami in case you prefer, we can restart the tests after resetting the parameters to the standard values before you sign the PR. Let me know what you prefer, or feel free to do it :)

**Comment by tvami** [comment]
>  we can restart the tests 

I dont think we need that, we can just use the backport tests to convince ourselves

**Comment by tvami** [comment]
+db
 * PR according to desc
 * tests pass, diffs are understood

**Comment by perrotta** [comment]
+1

**Pull request merged**